### ISSUE TITLE
Error: "Protocol not available"

### ISSUE THREAD

**Issue by Kixunil** [hyperium/hyper#1509]
When trying to run the server:

```rust
let core = tokio_core::reactor::Core::new().unwrap();

// I tried both "127.0.0.1:8331", "0.0.0.0:8331" and even "[::1]:8331", also other ports
let addr = std::net::SocketAddr::new(config.bind_address, config.bind_port);
// ProxyHandle implements Service
let service = ProxyHandle(Rc::new(Proxy { config, client: HttpClient::new(&core.handle()), }));

let server = hyper::server::Http::new().bind(&addr, move || Ok(service.clone())).unwrap();
server.run().unwrap();
```

It panics with OS error "Protocol not available"

`nc -l` works, so I guess the problem is somewhere in `hyper` or its dependency.

**Comment by seanmonstar** [comment]
Oh noes! What version of hyper are you using?

And when you say it panics, where? At one of the `unwrap`s you have, or inside hyper?

**Comment by Kixunil** [comment]
The last one - 0.11.26

It panics on my unwrap - the one at the end of line: `let server = hyper::server::Http::new().bind(&addr, move || Ok(service.clone())).unwrap();`

**Comment by Kixunil** [comment]
FYI, I just rewrote the code to use `tokio_core::net::TcpListener::bind(&addr, &handle).unwrap();` and `hyper::server::Http::new().serve_incoming(listener.incoming().map(|(socket, _addr)| socket), move || Ok(service.clone()));` and it works.

**Comment by sfackler** [comment]
I've seen the "protocol unavailable" message once before myself but I can't remember what the context was unfortunately.

**Comment by seanmonstar** [comment]
What OS is this happening on?

**Comment by Kixunil** [comment]
Linux Mint

**Comment by seanmonstar** [comment]
Do you what version of the Linux kernel it is? 0.11.26 tries to set `SO_REUSEPORT` and `SO_REUSEADDR`, so that the newer `run_threads` can work. Looks like `SO_REUSEPORT` is only in Linux >=3.9?

**Comment by asonix** [comment]
mint should definitely be at least 3.9, unless that's a really old install.

I was on 3.12 in 2013 on elementary OS (another ubuntu derivative)

**Comment by seanmonstar** [comment]
So, I think https://github.com/hyperium/hyper/commit/2c48101a6ee1269d7c94a0c3e606b2d635b20615 will fix this, if you want to try it out before I just publish a new version assuming it fixes it :)

**Comment by Kixunil** [comment]
I plan to try it. Thank you for quick response!

**Comment by seanmonstar** [comment]
I've published v0.11.27 with the fix mentioned here!
